coveo/ui-kit
# feat(atomic): add support for RGA rich formatting

# [SVCC-3606](https://coveord.atlassian.net/browse/SVCC-3606)

This PR adds support for RGA rich formatting to the Atomic Generated Answer component. Here is how it works:

1. `atomic-generated-answer` sets its capabilities using the `contentFormat` configuration parameter (e.g., `contentFormat: ['text/markdown', 'text/plain']`.
2. When reading the stream, the actual answer format is communicated to Headless via a header message. The content format is stored in the `generated-answer` state.
3. Upon rendering, the `atomic-generated-answer` component checks the content format. When receiving `text/markdown` a transformation pipeline is invoked to transform the output to HTML to be injected in the DOM.

The transformation is performed by [Marked](https://github.com/markedjs/marked). Some custom renderers have been added to handle the following:

* remove unnecessary paragraph elements when list items are separated by empty lines
* table elements are wrapped in a scrolling container to enforce a maximum height
* some inline elements such as bold, italic, and code are automatically closed to avoid showing Markdown artifacts when streaming the answer

![atomic-question-1](https://github.com/coveo/ui-kit/assets/10946843/e9f38b7c-bbf2-4138-a3a5-c42f00b8c190)

![atomic-question-2](https://github.com/coveo/ui-kit/assets/10946843/1efa801d-7179-4a54-bd92-48c91388fee9)

## How to test

1. You must have a working RGA model with rich formatting enabled
2. Select the Gen Q&A
3. When asking a question, the rich formatting should kick in when using the "Auto" answer style. Rich formatting is not yet supported for other answer styles.


[SVCC-3606]: https://coveord.atlassian.net/browse/SVCC-3606?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
